README.rst
django-dev
==========
https://github.com/idlesign/django-dev

.. image:: https://img.shields.io/pypi/v/django-dev.svg
    :target: https://pypi.python.org/pypi/django-dev

.. image:: https://img.shields.io/pypi/dm/django-dev.svg
    :target: https://pypi.python.org/pypi/django-dev

.. image:: https://img.shields.io/pypi/l/django-dev.svg
    :target: https://pypi.python.org/pypi/django-dev


Description
-----------

*Tools to facilitate application development for Django*

Console utility for those who have more than one Django reusable application to maintain.

It allows certain actions to be done in batch mode over all you apps.


Supported commands:

* **bootstrap**

* **add_migrations**

* **make_trans**

* **list_apps**

* **list_venvs**

* **install_package**



Requirements
------------

1. Python 2.7+, 3.2+
2. Python `virtualenv` package (or just Python 3.3+).


How to start
------------

1. Make a new directory, where you want your workspace to be, and step into it;

2. Run ``> django-dev bootstrap`` to create basic directory structure;

3. Symlink your apps into ``apps`` directory (was created by step 2 in your current directory).

    Example:

    ``ln -s /home/idle/dev/django-sitetree/sitetree /home/idle/dev/dj_workspace/apps/sitetree``

    Notice, that symlinked is a directory containing *models.py* (not *setup.py*).



Making both South and Django 1.7+ migrations
--------------------------------------------

Use **add_migrations** command.

* Relocate old South migrations into *south_migrations* and create new migrations in *migrations*:

  ``> django-dev add_migrations --relocate_south``

* Create/update migrations both for South and Django 1.7:

  ``> django-dev add_migrations``

* Create/update migrations for certain apps:

  ``> django-dev add_migrations --apps sitecats siteflags``



Updating translation files
--------------------------

Use **make_trans** command.

* Update existing .mo and .po files for every locale available in app:

  ``> django-dev make_trans``

* Update/create .mo and .po files for certain locales:

  ``> django-dev make_trans ru en``

* Update/create .mo and .po files for certain apps:

  ``> django-dev make_trans ru --apps sitetree sitegate``



Satisfy your app requirements
-----------------------------

Use **install_package** command to install packages into virtual envs to satisfy app requirements.

Exact version numbers could be given as in PIP: ``somedep==1.5``.



Get involved into django-dev
----------------------------

**Submit issues.** If you spotted something weird in application behavior or want to propose a feature you can do that at https://github.com/idlesign/django-dev/issues

**Write code.** If you are eager to participate in application development, fork it at https://github.com/idlesign/django-dev, write your code,
whether it should be a bugfix or a feature implementation, and make a pull request right from the forked project page.

**Spread the word.** If you have some tips and tricks or any other words in mind that you think might be of interest for the others — publish it.


CHANGELOG
django-dev changelog
====================


v0.2.1
------
* Fixed `unexpected keyword argument 'version'` for Py3.


v0.2.0
------
+ Added `install_package` command.
* Default Django version is bumped up to 1.8.


v0.1.4
------
* Fixed `add_migrations` with `--relocate_south` flag unable to delete `migrations` dir.


v0.1.3
------
* Fixed south install stdout redirect.
+ Added user-friendly error messages for console commands.


v0.1.2
------
* Implemented buggy migrations fix for those created with `makemigrations` using Py2.
* Added most Django contribs to installed apps into manage.py to satisfy 1.7 migrations proc.


v0.1.1
------
* Fixed false South initial migration add on --relocate_south.


v0.1.0
------
+ Basic functionality.


django_dev/dev.py
import argparse
import logging
import os
from shutil import copytree, rmtree
from subprocess import Popen, PIPE

try:
    from venv import create as create_venv
    VENV_CREATE_KWARGS = {'symlinks': True}
except ImportError:
    from virtualenv import create_environment as create_venv
    VENV_CREATE_KWARGS = {'symlink': True}

try:
    import configparser
except ImportError:
    import ConfigParser as configparser

from . import VERSION


DIR_CURRENT = os.getcwd()
DJANGO_DEFAULT_VERSION = '1.8'
APPS_DIRNAME = 'apps'
VENVS_DIRNAME = 'venvs'


MANAGE_PY = '''
# This file is created automatically by django-dev.
# Do not try to edit it. All changes manually done will be lost.
import os
import sys


if __name__ == '__main__':
    #PROJECT_PATH = os.path.realpath(os.path.dirname(__file__))
    sys.path = ['%(apps_path)s'] + sys.path

    try:
        import south
        south = ('south',)
    except:
        south = ()

    from django import get_version
    print('* Django: {}'.format(get_version()))

    from django.conf import settings, global_settings
    INSTALLED_APPS = (
        'django.contrib.admin',
        'django.contrib.auth',
        'django.contrib.contenttypes',
        'django.contrib.sessions',
        'django.contrib.messages',
        'django.contrib.staticfiles',
        'django.contrib.sites',
        'django.contrib.flatpages',
    )
    if not settings.configured:
        settings.configure(
            INSTALLED_APPS=INSTALLED_APPS + south + ('%(apps_available)s',),
            DATABASES={'default': {'ENGINE': 'django.db.backends.sqlite3'}},
            MIDDLEWARE_CLASSES=global_settings.MIDDLEWARE_CLASSES,  # Prevents Django 1.7 warning.
            SOUTH_MIGRATION_MODULES = {%(south_migration_modules)s}
        )

    try:  # Django 1.7 +
        from django import setup
        setup()
    except ImportError:
        pass


    from django.core.management import execute_from_command_line
    execute_from_command_line(sys.argv)
'''


class DjangoDevException(Exception):
    """Exception class used by django-dev."""


class DevTools(object):

    def __init__(self, log_level=None):
        self.workshop_path = DIR_CURRENT
        self.apps_path = os.path.join(self.workshop_path, APPS_DIRNAME)
        self.venvs_path = os.path.join(self.workshop_path, VENVS_DIRNAME)
        self.logger = logging.getLogger(self.__class__.__name__)
        self.configure_logging(log_level)

    def configure_logging(self, verbosity_lvl=None, format='%(message)s'):
        """Switches on logging at a given level.

        :param verbosity_lvl:
        :param format:

        """
        if not verbosity_lvl:
            verbosity_lvl = logging.INFO
        logging.basicConfig(format=format)
        self.logger.setLevel(verbosity_lvl)

    def _run_shell_command(self, command, pipe_it=True):
        """Runs the given shell command.

        :param command:
        :return: bool Status
        """
        stdout = None
        if pipe_it:
            stdout = PIPE
        self.logger.debug('Executing shell command: %s' % command)
        return not bool(Popen(command, shell=True, stdout=stdout).wait())

    def _get_venv_path(self, dj_version):
        """Returns virtual env directory path (named after Django version).

        :param str dj_version:
        :rtype: str
        :return: path
        """
        return os.path.join(self.venvs_path, dj_version)

    def _get_app_path(self, app_name):
        """Returns application directory path for a given app name.

        :param str app_name:
        :return: path
        """
        return os.path.join(self.apps_path, app_name)

    def _get_manage_py_path(self):
        """Returns a path to manage.py file.

        :rtype: str
        :return: path
        """
        return os.path.join(self.workshop_path, 'manage.py')

    def run_manage_command(self, command, venv_path, verbose=True):
        """Runs a given Django manage command in a given virtual environment.

        :param str command:
        :param str venv_path:
        :param bool verbose:
        """
        self.logger.debug('Running manage command `%s` for `%s` ...' % (command, venv_path))
        self._run_shell_command(
            '. %s/bin/activate && python %s %s' % (venv_path, self._get_manage_py_path(), command),
            pipe_it=(not verbose))

    def venv_install(self, package_name, venv_path):
        """Installs a given python package into a given virtual environment.

        :param str package_name:
        :param str venv_path:
        """
        self.logger.debug('Installing `%s` into `%s` ...' % (package_name, venv_path))
        self._run_shell_command('. %s/bin/activate && pip install -U %s' % (venv_path, package_name))

    def make_venv(self, dj_version):
        """Creates a virtual environment for a given Django version.

        :param str dj_version:
        :rtype: str
        :return: path to created virtual env
        """
        venv_path = self._get_venv_path(dj_version)
        self.logger.info('Creating virtual environment for Django %s ...' % dj_version)
        try:
            create_venv(venv_path, **VENV_CREATE_KWARGS)
        except ValueError:
            self.logger.warning('Virtual environment directory already exists. Skipped.')
        self.venv_install('django==%s' % dj_version, venv_path)
        return venv_path

    def make_apps_dir(self):
        """Creates an empty directory for symlinks to Django applications.

        :rtype: str
        :return: created directory path
        """
        self.logger.info('Creating a directory for symlinks to your Django applications `%s` ...' % self.apps_path)
        try:
            os.mkdir(self.apps_path)
        except OSError:
            pass  # Already exists.
        return self.apps_path

    def dispatch_op(self, op_name, args_dict):
        """Dispatches an operation requested.

        :param str op_name:
        :param dict args_dict:
        """
        self.logger.debug('Requested `%s` command with `%s` args.' % (op_name, args_dict))
        method = getattr(self, 'op_%s' % op_name, None)
        if method is None:
            error_str = '`%s` command is not supported.' % op_name
            self.logger.error(error_str)
            raise DjangoDevException(error_str)
        method(**args_dict)
        self.logger.info('Done.')

    def get_venvs(self):
        """Returns a list of names of available virtual environments.

        :raises: DjangoDevException on errors
        :rtype: list
        :return: list of names
        """

        def raise_():
            error_str = 'Virtual environments are not created. Please run `bootstrap` command.'
            self.logger.error(error_str)
            raise DjangoDevException(error_str)

        if not os.path.exists(self.venvs_path):
            raise_()
        venvs = os.listdir(self.venvs_path)
        if not venvs:
            raise_()

        venvs.sort()
        return venvs

    def get_apps(self, only=None):
        """Returns a list of names of available Django applications,
        Optionally filters it using `only`.

        :param list|None only: a list on apps names to to filter all available apps against
        :raises: DjangoDevException on errors
        :rtype: list
        :return: list of apps names
        """
        if not os.path.exists(self.apps_path):
            error_str = 'It seems that this directory does not contain django-dev project. ' \
                        'Use `bootstrap` command to create project in the current directory.'
            self.logger.error(error_str)
            raise DjangoDevException(error_str)

        apps = os.listdir(self.apps_path)

        if not apps:
            error_str = 'Applications directory is empty. ' \
                        'Please symlink your apps (and other apps that you apps depend upon) into %s' % self.apps_path
            self.logger.error(error_str)
            raise DjangoDevException(error_str)

        apps.sort()
        if only is None:
            self.create_manage_py(apps)
            return apps

        diff = set(only).difference(apps)
        if diff:
            error_str = 'The following apps are not found: `%s`.' % ('`, `'.join(diff))
            self.logger.error(error_str)
            raise DjangoDevException(error_str)

        self.create_manage_py(apps)

        return [name for name in apps if name in only]

    def create_manage_py(self, apps):
        """Creates manage.py file, with a given list of installed apps.

        :param list apps:
        """
        self.logger.debug('Creating manage.py ...')
        with open(self._get_manage_py_path(), mode='w') as f:
            south_migration_modules = []
            for app in apps:
                south_migration_modules.append("'%(app)s': '%(app)s.south_migrations'" % {'app': app})

            f.write(MANAGE_PY % {
                'apps_available': "', '".join(apps),
                'apps_path': self.apps_path,
                'south_migration_modules': ", ".join(south_migration_modules)


            })

    def op_list_venvs(self):
        """Prints out and returns a list of known virtual environments.

        :rtype: list
        :return: list of virtual environments
        """
        self.logger.info('Listing known virtual environments ...')
        venvs = self.get_venvs()
        for venv in venvs:
            self.logger.info('Found `%s`' % venv)
        else:
            self.logger.info('No virtual environments found in `%s` directory.' % VENVS_DIRNAME)
        return venvs

    def op_list_apps(self):
        """Prints out and returns a list of known applications.

        :rtype: list
        :return: list of applications
        """
        self.logger.info('Listing known applications ...')
        apps = self.get_apps()
        for app in apps:
            self.logger.info('Found `%s`' % app)
        else:
            self.logger.info('\nDONE. No applications found in `%s` directory.\n' % APPS_DIRNAME)
        return apps

    def op_bootstrap(self):
        """Bootstraps django-dev by creating required directory structure."""
        self.logger.info('Bootstrapping django-dev directory structure in current directory ...')
        self.make_venv(DJANGO_DEFAULT_VERSION)
        venv_path = self.make_venv('1.6.5')
        self.venv_install('south==1.0.1', venv_path)
        apps_dir = self.make_apps_dir()
        self.logger.info('Now you may symlink (ln -s) your apps '
                         '(and other apps that you apps depend upon) into %s' % apps_dir)

    def _make_dirs(self, path):
        """Creates every directory in path as needed. Fails silently.

        :param str path:
        :rtype: bool
        :return: True if no exceptions raised; otherwise - False.
        """
        try:
            os.makedirs(path)
            return True
        except OSError:  # Probably already exists.
            pass
        return False

    def op_install_package(self, names):
        """Install packages into virtual envs as to satisfy app requirements.

        Exact version numbers could be given as in PIP: somedep==1.5

        :param list names:

        """
        venvs = self.get_venvs()
        for venv in venvs:
            for name in names:
                self.venv_install(name, self._get_venv_path(venv))

    def op_make_trans(self, locales=None, apps=None):
        """Generates/updates localization (.po, .mo) files for applications.

        :param list|None locales: Locales to generate files for. If `None` all available locales in apps are updated.
        :param list|None apps: Target applications filter. If `None` all available apps are processed.
        """
        self.logger.info('Making translations ...')
        apps = self.get_apps(only=apps)
        self.get_venvs()  # Sentinel.
        venv_path = self._get_venv_path(DJANGO_DEFAULT_VERSION)
        if locales is None:
            locales = []

        for app_name in apps:
            self.logger.info('Application: %s' % app_name)
            app_path = self._get_app_path(app_name)
            locales_path = os.path.join(app_path, 'locale')

            if not locales and os.path.exists(locales_path):  # Getting all existing locales.
                locales = os.listdir(locales_path)

            for lang in locales:
                self.logger.info('Locale: %s' % lang)
                locale_path = os.path.join(locales_path, '%s/LC_MESSAGES' % lang)
                self._make_dirs(locale_path)
                old_wd = os.getcwd()
                os.chdir(app_path)
                self.run_manage_command('makemessages -l %s' % lang, venv_path)
                self.run_manage_command('compilemessages -l %s' % lang, venv_path)
                os.chdir(old_wd)

    def op_add_migrations(self, apps=None, relocate_south=False):
        self.logger.info('Making migrations ...')
        apps = self.get_apps(only=apps)
        venvs = self.get_venvs()  # Sentinel.

        default_venv_path = self._get_venv_path(DJANGO_DEFAULT_VERSION)

        def fix_migrations(path):
            """Fixes buggy migrations created with `makemigrations` using Py2.
            See: https://code.djangoproject.com/ticket/23455

            :param path:
            :return:
            """
            self.logger.debug('Fixing migrations ...')
            for file in os.listdir(path):
                if os.path.splitext(file)[1] == '.py':
                    with open(os.path.join(path, file), 'r+') as f:
                        contents = f.read()
                        f.seek(0)
                        f.write(contents.replace("=b'", "='"))
                        f.truncate()

        last_app_name = None
        for venv in venvs:
            venv_path = self._get_venv_path(venv)

            for app_name in apps:
                app_path = self._get_app_path(app_name)
                PATH_SOUTH = os.path.join(app_path, 'south_migrations')
                PATH_BUILTIN = os.path.join(app_path, 'migrations')
                south_exists = os.path.exists(PATH_SOUTH)

                if last_app_name != app_name:
                    self.logger.info('Application: %s' % app_name)
                    last_app_name = app_name

                if relocate_south and not south_exists and os.path.exists(PATH_BUILTIN):  # Foolproof.
                    self.logger.info('Relocating South migrations into %s ...' % PATH_SOUTH)
                    copytree(PATH_BUILTIN, PATH_SOUTH)
                    rmtree(PATH_BUILTIN)
                    south_exists = True

                # Add migrations for both.
                if venv_path == default_venv_path:  # Django with migrations built-in (1.7+)
                    self.run_manage_command('makemigrations %s' % app_name, venv_path)

                else:
                    flag = '--auto' if south_exists else '--initial'
                    self.run_manage_command('schemamigration %s %s' % (app_name, flag), venv_path)

                if os.path.exists(PATH_BUILTIN):
                    fix_migrations(PATH_BUILTIN)


def main():
    arg_parser = argparse.ArgumentParser(
        prog='django-dev', description='Tools to facilitate application development for Django')

    arg_parser.add_argument('--version', action='version', version='.'.join(map(str, VERSION)))
    arg_parser.add_argument('--debug', help='Show debug messages while processing', action='store_true')

    arg_parser_apps = argparse.ArgumentParser(add_help=False)
    arg_parser_apps.add_argument(
        '--apps', nargs='+', help='Whitespace-separated list of applications names. Example: sitecats, siteflags.')

    sub_parsers = arg_parser.add_subparsers(dest='subparser_name')
    sub_parsers.add_parser('bootstrap', help='Creates a basic django-dev directory structure in a current directory.')
    sub_parsers.add_parser('list_apps', help='Prints out currently available applications.')
    sub_parsers.add_parser('list_venvs', help='Prints out currently available virtual environments.')

    sub_parser_add_migrations = sub_parsers.add_parser(
        'add_migrations', help='Adds both South and Django 1.7+ migrations for apps.', parents=[arg_parser_apps])
    sub_parser_add_migrations.add_argument(
        '--relocate_south',
        help='Flag to relocate old South migrations from `migrations` into `south_migrations` folder.',
        action='store_true')

    sub_parser_make_trans = sub_parsers.add_parser(
        'make_trans', help='Creates translation (.po, .mo) files for the given locales.', parents=[arg_parser_apps])
    sub_parser_make_trans.add_argument(
        'locales', nargs='*',
        help='Locales identifiers to make localization files for. Whitespace-separated values are allowed. '
             'Example: ru en.')

    sub_parser_install_package = sub_parsers.add_parser(
        'install_package', help='Installs packages into virtual environments.')
    sub_parser_install_package.add_argument(
        'names', nargs='*',
        help='Package names to install. Exact version number could be supplied as for PIP: some_package==1.2.3 '
             'Whitespace-separated values are allowed.')

    parsed_args = arg_parser.parse_args()
    parsed_args = vars(parsed_args)  # Convert args to dict

    log_level = None
    if parsed_args['debug']:
        log_level = logging.DEBUG
    del parsed_args['debug']

    dt = DevTools(log_level=log_level)

    target_subparser = parsed_args['subparser_name']
    del parsed_args['subparser_name']

    try:
        dt.dispatch_op(target_subparser, parsed_args)
    except DjangoDevException:
        pass  # This will be logged in into stdout.
